drivers/hmis/app/graphql/types/hmis_schema/ce_referral.rb

```diff
@@ -17,9 +17,16 @@ class HmisSchema::CeReferral < Types::BaseObject
     def steps
       instance = object.workflow_instance
       steps_by_node_id = instance.steps.index_by(&:node_id)
-      instance.template.graph.walk.filter(&:task?).map do |node|
-        steps_by_node_id[node.id] || instance.steps.new(node: node).freeze
-      end
+
+      graph = instance.template.graph(preloads: :inflows) # preload inflows so we can check conditions without n+1
+      graph.
+        # Stop the search when the node doesn't exist yet and is conditional. We don't want to return this node, or any of its children, if it won't definitely happen.
+        walk(stop_when: ->(node) { steps_by_node_id[node.id].nil? && node.inflows.map(&:condition).any? }).
+        filter(&:task?).
+        map do |node|
+        # If this node exists already, return it; otherwise, return a non-persisted version, ONLY IF it has no conditionals (will definitely happen).
+        steps_by_node_id[node.id] || (node.inflows.map(&:condition).any? ? nil : instance.steps.new(node: node).freeze)
+      end.compact
```

Comment:
Looks good to me!
Suggestion for readability improvement, since this is pretty complicated:

```ruby
        walk(stop_when: ->(node) { steps_by_node_id[node.id].nil? && node.has_conditional_inflows? }).
        filter(&:task?).
        map do |node|
            return steps_by_node_id[node.id] if steps_by_node_id[node.id] # task instance already exists
            
            return nil if node.has_conditional_inflows? # task is conditional, don't show it yet
            
            # initialize step to display in the UI
            instance.steps.new(node: node).freeze
      end.compact
```